src/main/java/org/jenkinsci/plugins/workflow/support/actions/LogStorageAction.java

```diff
@@ -0,0 +1,159 @@
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import org.apache.commons.jelly.XMLOutput;
+import org.jenkinsci.plugins.workflow.actions.FlowNodeAction;
+import org.jenkinsci.plugins.workflow.actions.LogAction;
+import org.jenkinsci.plugins.workflow.actions.PersistentAction;
+import org.jenkinsci.plugins.workflow.flow.FlowExecutionOwner;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.log.LogStorage;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.DoNotUse;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * A marker for a node which had some log text using {@link LogStorage#nodeListener}.
+ */
+@Restricted(NoExternalUse.class) // for use from DefaultStepContext only
+public class LogStorageAction extends LogAction implements FlowNodeAction, PersistentAction {
+
+    private static final Logger LOGGER = Logger.getLogger(LogStorageAction.class.getName());
+
+    public transient FlowNode node;
+
+    private LogStorageAction(FlowNode node) {
+        if (!node.isActive()) {
+            throw new IllegalStateException("cannot start writing logs to a finished node " + node);
+        }
+        this.node = node;
+    }
+
+    @Override public void onLoad(FlowNode node) {
+        this.node = node;
+    }
+
+    @Override public AnnotatedLargeText<? extends FlowNode> getLogText() {
+        FlowExecutionOwner owner = node.getExecution().getOwner();
+        return LogStorage.of(owner).stepLog(node, !node.isActive());
+    }
+
+    /**
+     * Used from <tt>console.jelly</tt> to write annotated log to the given output.
+     */
+    @Restricted(DoNotUse.class) // Jelly
+    public void writeLogTo(long offset, XMLOutput out) throws IOException {
+        getLogText().writeHtmlTo(offset, out.asWriter());
+    }
+
+    /**
+     * Creates a sink to print output from a step.
+     * Will use {@link LogActionImpl} if necessary.
+     * @param node a node which wishes to print output
+     * @param filter an optional log filter (must be {@link Serializable})
+     * @return a stream
+     */
+    @SuppressWarnings("deprecation") // LogActionImpl here for backward compatibility
+    public static @Nonnull TaskListener listenerFor(@Nonnull FlowNode node, @CheckForNull ConsoleLogFilter filter) throws IOException, InterruptedException {
+        FlowExecutionOwner owner = node.getExecution().getOwner();
+        if (LogActionImpl.isOld(owner) || node.getAction(LogActionImpl.class) != null) {
+            return LogActionImpl.stream(node, filter);
+        } else {
+            if (node.getAction(LogStorageAction.class) == null) {
+                node.addAction(new LogStorageAction(node));
+            }
+            TaskListener listener = LogStorage.of(owner).nodeListener(node);
+            if (filter != null) {
+                if (filter instanceof Serializable) {
```

Comment:
They _must_ be `Serializable` in order to work with Pipeline at all, since otherwise the attempt to save context to `program.dat` would explode. This also applies to `SimpleBuildWrapper.createLoggerDecorator`.

This is just a best effort to deal with unserializable filters. An implementation of JENKINS-30777 would add global filters to the mix, though again this could simply exclude non`Serializable` ones from the start. `Serializable` is a marker interface, which is an antipattern in Java (IMHO), because it is impossible for an implementation such as `MergedFilter` cannot conditionally indicate serializability depending on its delegates. In this case it would not be worth much, since again _all_ Pipeline filters are expected to be `Serializable`.

Really what we want to know is not just whether the filter is `Serializable`, which merely means that its instance fields can be saved and restored, but whether it is _remotable_—whether its `decorateLogger` could meaningfully run from an agent JVM. We have no API for that currently. There is the marker interface from @oleg-nenashev called `SerializableOnlyOverRemoting`, but implementing that implies that the class _cannot_ be serialized to `program.dat`, which is not true—we want to assert that an object can be _both_ serialized normally and remoted. That would be a new Remoting and thus core API which plugins providing filters would need to update to. Probably overkill for something which can anyway be tested just by using new `workflow-*` test deps and running something on `JenkinsRule.createSlave`.